### PR TITLE
Open handler file when initializing Dotnet project

### PR DESCRIPTION
**Pull request by micholas**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Infrastructure (change which improves the lifecycle management (CI/CD, build, package, deploy, lint, etc) of the application, but does not change functionality.)
- [ ] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)
- [ ] Testing (change which modifies or adds test coverage, but does not change functionality)
- [ ] Documentation (change which modifies or adds documentation, but does not change functionality)

## Description

When a new C# project is initialized,  the handler file should be opened for editing.

## Related Issue(s)

https://github.com/aws/aws-toolkit-vscode/issues/288

## Testing

Tested `getMainSourceFileUri` for Dotnet projects

## Checklist

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/test/lambda/utilities/getMainSourceFile.test.ts [hunk 1]
```diff
@@ -317,7 +317,8 @@ describe('getMainSourceFile', async () => {
             await test('python')
             await test('python2.7')
             await test('python3.6')
-       })
+            await test('python3.7')
```
Comment (awschristou) [inline src/test/lambda/utilities/getMainSourceFile.test.ts]: Because we're working in a feature branch, could you please make this change in a separate PR aimed at the develop branch? It looks like something we wouldn't want to wait until the dotnet feature branch is complete.

File: src/test/lambda/utilities/getMainSourceFile.test.ts [hunk 2]
```diff
@@ -342,4 +343,81 @@ describe('getMainSourceFile', async () => {
...
+        })
+
+        it('recognizes all Dotnet runtimes', async () => {
+            async function test(runtime?: string): Promise<void> {
+                const templateUri = vscode.Uri.file(path.join('/dir', 'template.yaml'))
+                const actual: vscode.Uri = await getMainSourceFileUri({
+                    root: vscode.Uri.file('/dir'),
+                    getLocalTemplates: (...workspaceUris: vscode.Uri[]) => toAsyncIterable([templateUri]),
+                    loadSamTemplate: async uri => {
+                        const template = createTestTemplate()
```
Comment (awschristou) [inline src/test/lambda/utilities/getMainSourceFile.test.ts]: `createTestTemplate` could have a `runtime` arg that sets this, and defaults to `dotnet`

File: src/test/lambda/utilities/getMainSourceFile.test.ts [hunk 3]
```diff
@@ -342,4 +343,81 @@ describe('getMainSourceFile', async () => {
...
+                await getMainSourceFileUri({
+                    root: vscode.Uri.file('/dir'),
+                    getLocalTemplates: (...workspaceUris: vscode.Uri[]) => toAsyncIterable([
+                        vscode.Uri.file('/dir/template.yaml')
+                    ]),
+                    loadSamTemplate: async uri => createTestTemplate(),
+                    fileExists: async p => false
+                })
+            } catch (err) {
+                // const expectedBasePath = path.join('/dir', 'my_app', 'app')
```
Comment (awschristou) [inline src/test/lambda/utilities/getMainSourceFile.test.ts]: Remove commented code

File: src/test/lambda/utilities/getMainSourceFile.test.ts [hunk 4]
```diff
@@ -342,4 +343,81 @@ describe('getMainSourceFile', async () => {
...
+                    fileExists: async p => path.extname(p) === '.cs'
+                })
+
+                assert.strictEqual(actual.fsPath, expectedHandlerPath)
+            }
+
+            await test('dotnet')
+            await test('dotnetcore1.0')
+            await test('dotnetcore2.0')
+            await test('dotnetcore2.1')
```
Comment (awschristou) [inline src/test/lambda/utilities/getMainSourceFile.test.ts]: An issue with a test that internally performs multiple tests is that the whole test stops on first failure. Ideally you have feedback on each test's pass/fail, regardless of other test results. Please adjust to something [closer to parameterized tests](https://mochajs.org/#dynamically-generating-tests) - this is something we should be doing moving forward.

File: src/lambda/utilities/getMainSourceFile.ts [hunk 5]
```diff
@@ -146,3 +148,21 @@ async function getPythonSourceFileUri({
...
+    fileExists,
+    root,
+    resource,
+}: Pick<OpenMainSourceFileUriContext, 'fileExists'> & {
+    root: vscode.Uri,
+    resource: CloudFormation.Resource
+}): Promise<vscode.Uri> {
+    const handler = resource.Properties!.Handler
+    const [packageName] = handler.split('::')
+    const handlerFilePath = path.join(root.fsPath, 'src', packageName, 'Program.cs')
```
Comment (awschristou) [inline src/lambda/utilities/getMainSourceFile.ts]: I feel like this should be the "fallback" path, if no other file can be dynamically determined.
Given that codeuri can essentially point at a zip file "anywhere", all we have to work with is the handler. @mpiroc is it reasonable to try to pull the class name from the handler and try looking for a file with the same name? Failing this, it could then "fall back" to looking for Program.cs.

Comment (micholas) [reply src/lambda/utilities/getMainSourceFile.ts]: Correct me if I'm wrong but this is only used when creating a new SAM app and in this case the file name, code directory, namespace and handler are constant; i.e. the same thing is created every time SAM creates a Dotnet app. If is truly a constant, the code is already more complicated than need be and there is little point in spending more effort to further complicate it to support other "what if the code is elsewhere" scenarios.

It is apparent by the directory structure that there are broader objectives for `getMainSourceFileUri` but it is not used in broader scenarios and thus could be represented by an internal function within `createNewSamApp.ts`

File: src/test/lambda/utilities/getMainSourceFile.test.ts (hunk 4 above)
Comment (micholas) [reply src/test/lambda/utilities/getMainSourceFile.test.ts]: The tests for Nodejs, Python and Dotnet follow the same flow. Is your belief that these tests were never OK but that now would be a good time to refactor all of these tests?

Given that `getMainSourceFile` is only used to open the the handler source file after creating a new SAM app and that SAM always creates the same handler file... do you think that perhaps we could consider simplifying all of the is by replacing it with a map `{[runtimename: string]: handlerFilePath: string}`?

File: src/test/lambda/utilities/getMainSourceFile.test.ts (hunk 1 above)
Comment (micholas) [reply src/test/lambda/utilities/getMainSourceFile.test.ts]: Done. Removed `await test('python3.7')`.
✔️

File: src/test/lambda/utilities/getMainSourceFile.test.ts (hunk 2 above)
Comment (micholas) [reply src/test/lambda/utilities/getMainSourceFile.test.ts]: ✔️

File: src/test/lambda/utilities/getMainSourceFile.test.ts (hunk 3 above)
Comment (micholas) [reply src/test/lambda/utilities/getMainSourceFile.test.ts]: ✔️

File: src/test/lambda/utilities/getMainSourceFile.test.ts (hunk 4 above)
Comment (awschristou) [reply src/test/lambda/utilities/getMainSourceFile.test.ts]: It was an oversight in reviews until now. Ideally all of the tests would be made parameterized where it makes sense. Ideally each test verifies one concept/input/... . Leave it as-is if you like.

File: src/lambda/utilities/getMainSourceFile.ts (hunk 5 above)
Comment (awschristou) [reply src/lambda/utilities/getMainSourceFile.ts]: The SAM CLI is interchangeable, and there is no guarantee to the application it will produce. Otherwise it would make sense hardcode the expected file.

Comment (micholas) [reply src/lambda/utilities/getMainSourceFile.ts]: 1. Can you explain how the SAM CLI is interchangeable? I don't understand. I'm pretty confident you don't mean someone could roll their own SAM CLI compatible CLI.
2. Are you aware of any case in which the SAM CLI's `sam init` has generated a different handler for a specified runtime than it previously did?

It seems that a map along with an integration test that tests `sam init` for each runtime and verifies that the expected handler file exists would be much simpler to understand and maintain, more efficient, and no less robust than a case by case function for each runtime where the DotNet runtime considers all cases theoretically possible including with namespaces and without, and if a namespace is used, parsing each file in the directory associated with the namespace and attempting to identify the function corresponding to the handler in the SAM template.

To be specific. If a user or our extension calls `sam init --runtime dotnetcore2.1`, SAM _will_ create a file named `Program.cs` in `${samProjectDir}/src/HelloWorld` every time. We know how the movie ends, we don't need to run AI to consider all of the potential outcomes that could have been and prepare for each outcome, we can just deal with the known outcome. Or more specifically `sam init` can be viewed as  a pure function that always results in the same output for a given combination of inputs.

Testing this assumption simply means (in pseudo code):
```
for(const runtime in allSamRuntimes) {
   await exec(`sam init --runtime ${runtime}`)
   const relativeHandlerPath = relativeHandlerPathForRuntimes[runtime]
   const expectedHandlerExists = await fileExists(path.resolve(samProjectDir, 
relativHandlerPath)
   assert(expectedHandlerExists)
}
```

Yes, perhaps someday SAM could release a version of the CLI in which they decided to shuffle things around in which case our tests would fail against that SAM release and we'd need to modify our code check against a list of the known handler paths for a given runtime which is still simpler than the alternative.

Comment (micholas) [reply src/lambda/utilities/getMainSourceFile.ts]: If the team does decide to  continue to do things the more complicated way, we should consider leveraging the same logic as `getLambdaHandlerCandidates` that is being developed in the CR to support DotNet Run/Debug CodeLens. Basically we might list the `.cs` files in the directory associated with the namespace specified in the SAM template (assuming a namespace is used), run `getLambdaHandlerCandidates` on each and open for editing the first matching `.cs` file containing the function name specified in the SAM template.

Comment (micholas) [reply src/lambda/utilities/getMainSourceFile.ts]: But do keep in mind that no-one has yet to identify any situation in which the code in this CR does not work as intended so in some sense it appears any further changes are in the name of idealogical purity.... just another way of accomplishing the same thing. Unlike the run/debug code lens, during `sam init` we don't need to worry about every possible way a user could configure their template and code. SAM consistently initializes the way it initializes.

Comment (mpiroc) [reply src/lambda/utilities/getMainSourceFile.ts]: @micholas As an example, SAM CLI's .NET template currently creates a file called `Program.cs` that contains a class called `Function`, which breaks two .NET conventions: 1. Filename should match the class contained in the file; 2. `Program.cs` should contain the `Main` method, but the project generated by `sam init` is a class library (i.e., no `Main` method).

If SAM CLI changes the filename, do we want to force users to upgrade the toolkit? We only bump the minimum SAM version based on breaking changes, and a change to a template shouldn't be breaking.

Comment (micholas) [reply src/lambda/utilities/getMainSourceFile.ts]: @mpiroc To answer your question:

> If SAM CLI changes the filename, do we want to force users to upgrade the toolkit?

I believe the toolkit is already written such that it will warn users if they are using a version of the SAM CLI that is older or newer than was supported at the time the toolkit was generated. There are a host of reasons that it would likely be preferable for us to have users update the toolkit when a new version of SAM is released given the tightly coupled nature of the two. The unlikely chance that the SAM team would decide to generate completely different artifacts for previously supported runtimes could be just one more reason to synchronize SAM updates with toolkit updates.

I'm aware of the complexity of trying to cover every use case of trying to figure out where the handler generated by `sam init` might someday be or could have been (but isn't). I'm also very familiar with the handler functions that are in fact created. If I had implemented this for Java we wouldn't be having this conversation because both the directory and file name could be directly derived from the SAM template and I would have done so.

When `sam init` generates a DotNet project, the generated SAM template contains no reference to the actual name of the handler file.  This is why I brought up the idea of whether it is worth the effort of trying to algorithmically determine something we know in advance or whether we should consider simply using a map. **R2** question 2 would be one question we should definitely answer before choosing this simpler path. We may also consider asking the SAM team whether they have changed the path to a handler for a specific handler runtime, whether they intend to do so and what the likelihood is that they ever would. Given the response from most of our team members it appears the simplified approach is unpalatable regardless of the answer to these questions.

_Again if completely opposed to the idea of the simplified map approach, the other solution I mentioned in **R3** seems like the preferred option: inspecting all of the C# files in the directory associated with the handler namespace to determine which if any of them contain a function that might be the intended handler._

**Comment by micholas** [comment]
Replaced by https://github.com/aws/aws-toolkit-vscode/pull/481
